src/switchImplIntf.ml

```diff
@@ -0,0 +1,66 @@
+open Import
+
+let switch client document : string array Promise.t =
+  LanguageClient.sendRequest client ~meth:"ocamllsp/switchImplIntf"
+    ~data:(Uri.toString document.TextDocument.uri ~skipEncoding:true ())
+    ()
+
+let requestSwitch ~client ~capabilities document =
+  (* given a file uri, opens the file if it exists;
+     otherwise, creates the file but doesn't write it to disk *)
+  let showFile targetFileName =
+    let open Promise.O in
+    let uri = Uri.parse targetFileName in
+    Vscode.Workspace.openTextDocument (`Uri uri)
+    |> Js.Promise.catch (fun _ ->
+           (* if file does not exist *)
+           let createFileUri =
+             Uri.with_ uri @@ Uri.make_change ~scheme:"untitled" ()
+           in
+           Vscode.Workspace.openTextDocument (`Uri createFileUri))
+    >>= fun doc ->
+    Vscode.Window.showTextDocument (`Document doc) >>| fun _ -> ()
+  in
+
+  let filesToSwitchTo =
+    match (client, capabilities) with
+    | Some client, Some capabilities
+      when OcamlLsp.handleSwitchImplIntf capabilities ->
+      switch client document
+    | _ -> failwith "NOT IMPLEMENTED"
+    (* TODO handle more gracefully *)
+  in
+
+  let open Promise.O in
+  filesToSwitchTo >>= fun arr ->
+  match Array.to_list arr with
+  | [] ->
+    failwith
+      "all files must be mapped (can be switched to) to at least one file"
```

Comment:
Is this case possible at all?

Reply:
No, see the spec for the command [link](https://github.com/ulugbekna/ocaml-lsp/blob/5ba658835e2171da19d5053cca814075a0e79d13/ocaml-lsp-server/docs/ocamllsp/switchImplIntf-spec.md). How could I make this invariant more declarative in the code? I could put a comment saying such a case is impossible and `assert(false)` for example.

Reply:
Yes, that is preferable. `assert false` with an explanation sounds good to me.